### PR TITLE
[SPARK-39105][SQL] Add ConditionalExpression trait

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

Add `ConditionalExpression` trait.

### Why are the changes needed?

For developers, if a custom conditional like expression contains common sub expression then the evaluation order may be changed since Spark will pull out and eval the common sub expressions first during execution.

Add ConditionalExpression trait is friendly for developers.

### Does this PR introduce _any_ user-facing change?

no, add a new trait

### How was this patch tested?

Pass existed test

### CODE REVIEW

**Comment by ulysses-you** [comment]
cc @cloud-fan @viirya

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 1]
```diff
@@ -454,6 +454,33 @@ trait Nondeterministic extends Expression {
   protected def evalInternal(input: InternalRow): Any
 }
 
+/**
+ * An expression that contains predicate expression branch, so not all branch will be hit
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:458]:
```suggestion
 * An expression that contains conditional expression branches, so not all branches will be hit
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 2]
```diff
@@ -454,6 +454,33 @@ trait Nondeterministic extends Expression {
...
+   *        expression. We should only return the predicate expression.
+   * 2. CaseWhen: like `If`, the children of `CaseWhen` only get accessed in a certain
+   *              condition. We should only return the first condition expression as it
+   *              will always get accessed.
+   * 3. Coalesce: it's also a conditional expression, we should only return the first child,
+   *              because others may not get accessed.
+   * 4. NaNvl: we can only guarantee the left child can be always accessed.
+   *           And if we hit the left child, the right will not be accessed.
+   */
+  def head: Expression = children.head
```
[the remark this answers is not in the document]
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:475]: +1 for `Seq[Expression]` for 1.

[the remark this answers is not in the document]
Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: good call, Seq[Expression] is also good to me

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala [hunk 3]
```diff
@@ -127,56 +127,18 @@ class EquivalentExpressions {
...
-  //          false expressions in `If` may not get accessed, according to the predicate
-  //          expression. We should only recurse into the predicate expression.
-  //   3. CaseWhen: like `If`, the children of `CaseWhen` only get accessed in a certain
-  //                condition. We should only recurse into the first condition expression as it
-  //                will always get accessed.
-  //   4. Coalesce: it's also a conditional expression, we should only recurse into the first
-  //                children, because others may not get accessed.
-  //   5. NaNvl: it's a conditional expression, we can only guarantee the left child can be always
-  //             accessed. And if we hit the left child, the right will not be accessed.
+  //   2. ConditionalExpression: use it's specified expression
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala:130]:
```suggestion
  //   2. ConditionalExpression: use its children that will always be evaluated.
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 4]
```diff
@@ -454,6 +454,23 @@ trait Nondeterministic extends Expression {
   protected def evalInternal(input: InternalRow): Any
 }
 
+/**
+ * An expression that contains conditional expression branches, so not all branches will be hit.
+ * All optimization should be careful with the evaluation order.
+ */
+trait ConditionalExpression extends Expression {
+  /**
+   * Return the expression which can always be hit at runtime.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:463]:
```suggestion
   * Return the children expressions which can always be hit at runtime.
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala [hunk 5]
```diff
@@ -48,6 +48,14 @@ case class If(predicate: Expression, trueValue: Expression, falseValue: Expressi
   override def second: Expression = trueValue
   override def third: Expression = falseValue
   override def nullable: Boolean = trueValue.nullable || falseValue.nullable
+  /**
+   * Common subexpressions will always be evaluated at the beginning, but the true and
+   * false expressions in `If` may not get accessed, according to the predicate expression.
+   * We should only return the predicate expression.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala:54]: I think we can simplify the comment now
```
Only the condition expression will always be evaluated.
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala [hunk 6]
```diff
@@ -179,6 +187,36 @@ case class CaseWhen(
...
+    // We look at subexpressions in conditions and values of `CaseWhen` separately. It is
+    // because a subexpression in conditions will be run no matter which condition is matched
+    // if it is shared among conditions, but it doesn't need to be shared in values. Similarly,
+    // a subexpression among values doesn't need to be in conditions because no matter which
+    // condition is true, it will be evaluated.
+    val conditions = if (branches.length > 1) {
+      branches.map(_._1)
+    } else {
+      // If there is only one branch, the first condition is already covered by
+      // `head` and we should exclude it here.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala:206]:
```suggestion
      // `alwaysEvaluatedInputs` and we should exclude it here.
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala [hunk 7]
```diff
@@ -66,6 +67,19 @@ case class Coalesce(children: Seq[Expression]) extends ComplexTypeMergingExpress
...
   }
 
+  /**
+   * We should only return the first child, because others may not get accessed.
+   */
+  override def alwaysEvaluatedInputs: Seq[Expression] = children.head :: Nil
+
+  override def branchGroups: Seq[Seq[Expression]] = if (children.length > 1) {
+    // If there is only one child, the first child is already covered by
+    // `head` and we should exclude it here.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala:77]:
```suggestion
    // `alwaysEvaluatedInputs` and we should exclude it here.
```

**Comment by cloud-fan** [comment]
thanks, merging to master/3.3! (I'm backporting this small refactor as we will have a bug fix that relies on it)

**Comment by ulysses-you** [comment]
thank you @cloud-fan @viirya

**Comment by dongjoon-hyun** [comment]
+1 for @cloud-fan 's backporting decision.

Also, cc @MaxGekk

**Pull request closed without merging**